usr/sbin/pkcstok_migrate/pkcstok_migrate.c

```diff
@@ -0,0 +1,2367 @@
+    clear = malloc(clear_len);
+    if (cipher == NULL || clear == NULL) {
+        ret = CKR_HOST_MEMORY;
+        goto done;
+    }
+
+    rc = fread(cipher, cipher_len, 1, fp);
+    if (rc != 1) {
+        TRACE_ERROR("Cannot read %ld bytes from %s\n", cipher_len, mkfile);
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* decrypt the masterkey */
+    ret = compute_md5((char *)pin, strlen(pin), pin_md5_hash);
+    if (ret) {
+        TRACE_ERROR("Error calculating MD5 of PIN, ret=%08lX\n",ret);
+        goto done;
+    }
+
+    memcpy(des3_key, pin_md5_hash, MD5_HASH_SIZE);
+    memcpy(des3_key + MD5_HASH_SIZE, pin_md5_hash, DES_KEY_SIZE);
+
+    ret = sw_des3_cbc_decrypt(cipher, cipher_len, clear,
+                              &clear_len, (unsigned char *) "12345678",
+                              des3_key);
+    if (ret != CKR_OK) {
+        TRACE_ERROR("failed to decrypt master key file after read, ret=%08lX\n",ret);
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* compare the hashes to verify integrity */
+    ret = compute_sha1((char *)clear, MASTER_KEY_SIZE, hash_sha);
+    if (ret) {
+        TRACE_ERROR("Failed to compute sha1 for masterkey, ret=%08lX\n",ret);
+        goto done;
+    }
+
+    if (memcmp(hash_sha, clear + MASTER_KEY_SIZE, SHA1_HASH_SIZE) != 0) {
+        TRACE_ERROR("%s appears to be tampered! Cannot migrate.\n", mkfile);
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    memcpy(masterkey, clear, MASTER_KEY_SIZE);
+    ret = 0;
+
+done:
+    if (fp)
+        fclose(fp);
+    free(clear);
+    free(cipher);
+
+    return ret;
+}
+
+/**
+ * gets the stdll name from the given opencryptoki.conf file.
+ */
+CK_RV get_stdll_name(FILE *fp, char **stdll_name)
+{
+    char line[80], parm[80];
+    char *tmp_name;
+
+    snprintf(parm, sizeof(parm), "stdll =");
+    while (fgets(line, sizeof(line), fp)) {
+        if (strstr(line, "stdll =")) {
+            tmp_name = malloc(80);
+            if (!tmp_name)
+                return CKR_HOST_MEMORY;
+            sscanf(line, "stdll = %s", tmp_name);
+            *stdll_name = tmp_name;
+            return CKR_OK;
+        }
+    }
+
+    return CKR_FUNCTION_FAILED;
+}
+
+/**
+ * Check if the given conf_dir exists and contains the opencryptoki.conf.
+ */
+CK_BBOOL conffile_exists(const char *conf_dir)
+{
+    char fname[PATH_MAX];
+    struct stat statbuf;
+    DIR *dir;
+
+    dir = opendir(conf_dir);
+    if (dir == NULL) {
+        TRACE_INFO("Cannot open %s.\n", conf_dir);
+        return CK_FALSE;
+    }
+
+    /* Check if opencryptoki.conf exists */
+    memset(fname, 0, PATH_MAX);
+    snprintf(fname, PATH_MAX, "%s/opencryptoki.conf", conf_dir);
+    if (stat(fname, &statbuf) != 0) {
+        TRACE_INFO("Cannot find %s.\n", fname);
+        return CK_FALSE;
+    }
+
+    return CK_TRUE;
+}
+
+/**
+ * Check if the given data_store directory exists.
+ */
+CK_BBOOL datastore_exists(const char *data_store)
```

Comment:
those functions could be static. I would not make any functions static that are candidates to be moved to a common source file at a later time. But purely local helper functions can be static.

Reply:
ok